### PR TITLE
planner: rewrite FTS predicates to LIKE for evaluation of non-TiCI query plan

### LINKED ISSUE #68153: planner: query fails with MySQL compatible FTS search
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
CREATE TABLE bls_occupations (
      id INT PRIMARY KEY,                                                                                                                                                           
      title VARCHAR(200),                                                                                                                                                           
      description TEXT                                                                                                                                                              
    );                                                                                                                                                                              
                                                                                                                                                                         
INSERT INTO bls_occupations VALUES                                                                                                                                              
(1, 'Software Engineer',       'Designs, develops, and maintains software systems and applications.'),
(2, 'Senior Software Engineer','Leads design and architecture of complex distributed systems.'),                                                                               
(3, 'Database Administrator',  'Maintains relational databases and tunes query performance.'),                                                                                 
(4, 'Data Engineer',            'Builds data pipelines for analytics and machine learning workloads.'),
(5, 'Hardware Engineer',        'Designs printed circuit boards and embedded firmware.'),                                                                                      
(6, 'Civil Engineer',           'Designs roads, bridges, and other public infrastructure.'),                                                                                   
(7, 'Marketing Manager',        'Plans advertising campaigns and brand strategy.'),                                                                                            
(8, 'Technical Writer',         'Writes user manuals, API references, and tutorials.'),                                                                                        
(9, 'Security Analyst',         'Monitors networks for intrusions and audits access controls.'),                                                                               
(10,'Product Manager',          'Defines product roadmap and prioritizes engineering work.');    
```
### 2. What did you expect to see? (Required)

Successful execution.

### 3. What did you see instead (Required)
```
tidb>   SELECT id, title FROM bls_occupations   WHERE MATCH(title) AGAINST('Engin*' IN BOOLEAN MODE);
ERROR 1105 (HY000): cannot use 'MATCH ... AGAINST' outside of fulltext index
```

### 4. What is your TiDB version? (Required)

Master, 8.5 & X

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by terry1purcell**
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #68153

Problem Summary:

### What changed and how does it work?

### Summary                                                                                                                                                                         
                                                                                                                                                                                
When tidb_opt_enable_alternative_logical_plans=ON, adds a fallback that rewrites MATCH ... AGAINST to case-insensitive ILIKE predicates so full-text-search queries can execute without an FTS index. The rewrite is intentionally conservative: it only fires for a strict subset of search strings in direct-boolean predicate positions, and reaches the plan only when the round-1 native path is not viable. Anything outside that envelope either keeps the native builtin (errors at execution without an FTS index) or is rejected at plan time, never silently producing wrong rows.                                                                                                                                 
                                                                                                                                                                                
### Architecture

Round 1 (default, matches Alt-disabled behavior) — emits the native FTSMysqlMatchAgainst builtin. The expression rewriter records nonViableFTSMatch on the PlanBuilder when a direct-boolean-context MATCH cannot be served natively (no FTS index on a TiFlash replica, or a non-pushdown-safe modifier).
                                                                                                                                                                                  
Round 2: fts-like-fallback — fires only when round 1 reported a non-viable MATCH. The driver discards round 1's plan and re-runs the build with AlternativeLogicalPlanFTSLikeFallback=true, which switches the rewriter to ILIKE in direct-boolean positions. If this round also errors (e.g. unsupported search string with no FTS-index rescue), lastAltRoundErr propagates the message instead of the generic "failed to build logical plan" sentinel.                                                       
                                                                                                                                                                                
A single flag (AlternativeLogicalPlanFTSLikeFallback) drives the dispatch; viability state stays local to the build on PlanBuilder.nonViableFTSMatch.                           
   
### Where the rewrite applies                                                                                                                                                       
                                                                                                                                                                                
inDirectMatchBooleanContext (modeled on the existing canTreatInSubqueryAsExistsForFilter) walks the AST ancestor stack and accepts only:                                        
   
  - Root of WHERE / HAVING / JOIN ON, with ancestors limited to AND, OR, NOT, parentheses.                                                                                        
                                                                                                                                                                                
Everything else — MATCH ... > 0.5, MATCH ... = 0, MATCH ... IS NULL, MATCH inside CASE WHEN, arithmetic, scoring (SELECT field list, ORDER BY), etc. — keeps the native builtin, which preserves the float relevance score and errors at execution if no FTS index exists. Substituting a 0/1 integer in those positions would silently corrupt the comparison or sort.                                                                                                                                                                        
                                                                                                                                                                                
### Strict search-string subset

ValidateFTSSearchStringForLikeFallback rejects anything that would tokenize differently in MySQL FTS than a substring ILIKE match. Accepted by mode:                            
   
  - Natural-language mode: whitespace-separated alphanumeric word tokens only.                                                                                                    
  - Boolean mode: each token is word, +word, or -word, where word is alphanumeric (ASCII or non-ASCII UTF-8).                                                                   
                                                                                                                                                                                  
Rejected at plan time with error 1235 (ErrNotSupportedYet): phrases "...", prefix wildcard term*, relevance modifiers > < ~, grouping (...), mid-word punctuation like xx-yy, and any token containing %, _, \, ,, ., :, etc. WITH QUERY EXPANSION is likewise rejected (no ILIKE approximation exists).                                                      
                                                                                                                                                                                  
### Modifier handling                                                                                                                                                               
   
The tipb pushdown protocol does not serialize the FTS modifier (see distsql_builtin.go), so a Boolean-mode or WITH QUERY EXPANSION MATCH pushed down to TiFlash would silently execute as natural-language mode. To prevent this, matchAgainstToBuiltin rejects non-default modifiers at plan time unless matchHasLikeFallbackRescue is true (alt enabled + direct-boolean context, where the alt-rounds driver will discard the native plan and rebuild via fts-like-fallback). In practice:                                               
                                                                                                                                                                                
``` ┌──────────────────┬──────────────────────────────────┬─────────────────────────┐                                                                                               
  │     Modifier     │     Direct-boolean predicate     │ Scalar/scoring position │
  ├──────────────────┼──────────────────────────────────┼─────────────────────────┤                                                                                               
  │ Natural language │ native (round 1) or ILIKE rescue │ native                  │                                                                                             
  ├──────────────────┼──────────────────────────────────┼─────────────────────────┤
  │ Boolean          │ ILIKE rescue                     │ error 1235              │                                                                                               
  ├──────────────────┼──────────────────────────────────┼─────────────────────────┤                                                                                               
  │ With QE          │ error 1235 from ILIKE round      │ error 1235              │                                                                                               
  └──────────────────┴──────────────────────────────────┴─────────────────────────┘                                                                                               
 ```                                                                                                                                                                               

### NULL search handling                                                                                                                                                            
                                                                                                                                                                                
MATCH(c) AGAINST(NULL) matches nothing in MySQL FTS semantics, but three-valued logic matters under NOT: native evalReal returns NULL, so NOT NULL = NULL filters the row. The rewrite emits Constant(NULL) rather than Constant(0) so the same semantics hold under NOT, IS NULL, etc. The plan-cache skip is set before the NULL fast-path, so a prepared statement bound to NULL first followed by a non-NULL bind re-plans and returns correct rows instead of reusing a cached constant-false plan.                                    
                                                                                                                                                                                
### Boolean-mode operator support (post-strict-subset)                                                                                                                              
   
``` ┌───────────────┬─────────────────────────────────────────────────────────────────────┐                                                                                         
  │   Operator    │                              Behavior                               │                                                                                       
  ├───────────────┼─────────────────────────────────────────────────────────────────────┤                                                                                         
  │ +term         │ All required terms must match (AND across the per-term column DNFs) │
  ├───────────────┼─────────────────────────────────────────────────────────────────────┤                                                                                         
  │ -term         │ NOT(any column matches term)                                        │                                                                                         
  ├───────────────┼─────────────────────────────────────────────────────────────────────┤
  │ term          │ Optional — anchors the result only when no +term exists             │                                                                                         
  ├───────────────┼─────────────────────────────────────────────────────────────────────┤                                                                                         
  │ --only        │ Returns empty (matches MySQL)                                       │
  ├───────────────┼─────────────────────────────────────────────────────────────────────┤                                                                                         
  │ Anything else │ Rejected by ValidateFTSSearchStringForLikeFallback (1235)           │                                                                                       
  └───────────────┴─────────────────────────────────────────────────────────────────────┘                                                                                         
```   

### Plan cache                                                                                                                                                                      
                                                                                                                                                                                
  Marks the plan non-cacheable when the AGAINST argument is mutable across executions (? parameter marker, user variable, deferred expression). Literal AGAINST values keep the plan cacheable. The skip runs before the NULL fast-path so a NULL first bind can't bake a constant plan that gets reused later.
                                                                                                                                                                                  
### Selectivity                                                                                                                                                                   

BuildFTSToILikeExpressionFromBuiltin substitutes the equivalent ILIKE form for the opaque FTSMysqlMatchAgainst builtin so the round 1 native plan's cost reflects column histogram/TopN rather than the flat SelectivityFactor (0.8). Restricted to single-column MATCH because GetSelectivityByFilter declines multi-column expressions — a multi-column substitute would fall through to the same str-match default anyway.                                                                                                            
                                                                                                                                                                                
### Known semantic differences                                                                                                                                                      
   
These apply to ILIKE round queries only; the native path preserves full MySQL semantics:                                                                                                                                                                                                                                          
  - No relevance scoring (0/1 boolean output).                                                                                                                                    
  - No word boundaries (%cat% matches concatenate).                                                                                                                             
  - No stop-word filtering or minimum word length.
  - Optional-term relevance ranking is approximated as a filter when no required terms exist.                                                                                     
                                                                                                                                                                                  
### Files changed                                                                                                                                                                   
                                                                                                                                                                                  
```  ┌───────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
  │                           File                            │                                                    Purpose                                                     │
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤  
  │ pkg/expression/fts_to_like.go                             │ Strict-subset validator, ILIKE expression builder, mode-dispatch helpers, single-column substitution wrapper   │
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤  
  │ pkg/expression/fts_to_like_test.go                        │ Validator coverage (33 cases), BuildFTSToILikeExpressionFromBuiltin coverage                                   │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤  
  │ pkg/expression/builtin_fts.go                             │ Native builtinFtsMysqlMatchAgainstSig with modifier, defensive evalReal returning NULL for NULL search         │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤  
  │ pkg/planner/core/expression_rewriter.go                   │ matchAgainstToExpression dispatcher, inDirectMatchBooleanContext, ftsNativeViable, modifier guard,             │
  │                                                           │ matchHasLikeFallbackRescue                                                                                     │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/planner/core/fulltext_to_like.go                      │ Thin wrapper delegating to pkg/expression                                                                      │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/planner/core/fulltext_to_like_test.go                 │ Unit tests for ftsModifierAllowsNativePushdown, tableHasPublicFTSIndexOnColumn                                 │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/planner/core/planbuilder.go                           │ PlanBuilder.nonViableFTSMatch + HasNonViableFTSMatch/MarkNonViableFTSMatch accessors                           │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/planner/optimize.go                                   │ fts-like-fallback alternative round, round-1 plan invalidation, lastAltRoundErr                                │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/planner/cardinality/selectivity.go                    │ Selectivity substitution for FTSMysqlMatchAgainst (single-column only)                                         │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/sessionctx/stmtctx/stmtctx.go                         │ Single AlternativeLogicalPlanFTSLikeFallback flag                                                              │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ pkg/parser/ast/functions.go                               │ FTSMysqlMatchAgainst = "match_against" constant                                                                │  
  ├───────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ tests/integrationtest/t/planner/core/fulltext_search.test │ 94 integration test cases                                                                                      │  
  └───────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```                                                                                                                                                                                  
### Test plan                                                                                                                                                                     

  - 94 integration cases covering: natural-language mode, boolean mode (+/-/optional), empty/whitespace/NULL search, non-string columns, HAVING/JOIN ON predicates, NOT MATCH,    
  parenthesized MATCH, scalar positions (IS NULL, > 0.5, = 0, CASE), non-default modifiers in scoring/scalar/alt-disabled contexts, prepared statements (literal cacheable,
  parameter-marker non-cacheable, NULL-first-bind re-plans).                                                                                                                      
  - Unit: validator (33 cases), BuildFTSToILikeExpressionFromBuiltin (nil/wrong-fn/single-column/multi-column/NULL/non-subset), parser tests aligned with the strict subset.    
  - Strict-subset rejections previously-passing edge cases now expect error 1235.                                                                                                 
  - make lint passes.                                                                                                                                                             
  - gofmt clean.                                                                                            
                                                                                                                                                                                  
  🤖 Generated with https://claude.com/claude-code    


### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * LIKE-based fallback for MATCH...AGAINST so full-text queries run when native FTS isn’t viable; supports natural and boolean modes, preserves NULL semantics, and surfaces clear errors for unsupported patterns/modifiers.
  * Session controls to choose fallback vs. error behavior and planner handling across native/fallback rounds.
  * Improved planner handling and selectivity estimation for full-text predicates.

* **Tests**
  * Extensive unit and integration tests covering parsing, LIKE conversion, planner rewrites, prepared-plan cache behavior, and edge cases.

* **Chores**
  * Updated ignore rules and test sharding configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: pkg/sessionctx/variable/session.go [hunk 1]
```diff
@@ -1019,6 +1019,9 @@ type SessionVars struct {
 
 	// EnableCorrelationAdjustment is used to indicate if correlation adjustment is enabled.
 	EnableCorrelationAdjustment bool
+	// FulltextSearchFallback controls the behavior when MATCH...AGAINST syntax is used.
+	// Options: 'like' (convert to LIKE predicates), 'error' (throw error if no fulltext index).
```
[the remark this answers is not in the document]
Comment (terry1purcell) [reply pkg/sessionctx/variable/session.go]: unnecessary

**Comment by IANTHEREAL** [comment]
@pantheon-bot Please take a look at the PR and do a comprehensive bug hunt.

File: pkg/expression/integration_test/integration_test.go [hunk 2]
```diff
@@ -207,7 +207,12 @@ func TestFTSSyntax(t *testing.T) {
 	// tk.MustContainErrMsg("select * from t where (fts_match_word('hello', title)) > 0", "Currently 'FTS_MATCH_WORD()' must be used alone")
 	// tk.MustContainErrMsg("select (fts_match_word('hello', title)) AS score from t where fts_match_word('hello', title)", "Currently 'FTS_MATCH_WORD()' cannot be used in SELECT fields")
 	tk.MustContainErrMsg("select * from t where match() against ('hello')", `You have an error in your SQL syntax`)
-	tk.MustContainErrMsg("select * from t where match(title) against ('hello' in boolean mode)", `UnknownType: *ast.MatchAgainst`)
+	// Test MATCH...AGAINST with default 'like' fallback mode - should succeed
+	tk.MustQuery("select * from t where match(title) against ('hello' in boolean mode)")
+	// Test MATCH...AGAINST with 'error' fallback mode - should fail
+	tk.MustExec("set @@tidb_opt_fulltext_search_fallback='error'")
+	tk.MustContainErrMsg("select * from t where match(title) against ('hello' in boolean mode)", `fulltext search not supported`)
```
[the remark this answers is not in the document]
Comment (terry1purcell) [reply pkg/expression/integration_test/integration_test.go]: Old comment. No longer relevant

File: pkg/planner/core/expression_rewriter.go [hunk 3]
```diff
@@ -2252,6 +2254,79 @@ func (er *expressionRewriter) patternLikeOrIlikeToExpression(v *ast.PatternLikeO
...
+	}
+
+	if fallbackMode == "error" {
+		er.err = expression.ErrNotSupportedYet.GenWithStackByArgs("MATCH...AGAINST (fulltext search not supported, set tidb_opt_fulltext_search_fallback='like' to use LIKE fallback)")
+		return
+	}
+
+	// Fallback mode is "like" - convert to LIKE predicates
+	// Both the column expressions and Against expression have been visited
+	// and pushed onto the ctxStack. The stack layout is:
```
[the remark this answers is not in the document]
Comment (terry1purcell) [reply pkg/planner/core/expression_rewriter.go]: No longer relevant

[the remark this answers is not in the document]
Comment (terry1purcell) [reply pkg/planner/core/expression_rewriter.go]: This comment deviates from the MySQL semantics. Ignoring this comment.

File: pkg/planner/core/fulltext_to_like.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+		// AND all required/excluded predicates together
+		if len(allPredicates) == 0 {
+			return &expression.Constant{
+				Value:   types.NewIntDatum(0),
+				RetType: types.NewFieldType(mysql.TypeTiny),
+			}, nil
+		}
+
+		return expression.ComposeCNFCondition(er.sctx, allPredicates...), nil
```
[the remark this answers is not in the document]
Comment (terry1purcell) [reply pkg/planner/core/fulltext_to_like.go]: No longer relevant

**Comment by terry1purcell** [comment]
@pantheon-bot please review

File: pkg/expression/fts_to_like.go [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+//
+// Semantic differences from MySQL's full-text search are documented in detail
+// at the planner-level call site; this helper preserves those approximations
+// so both callers see the same translated expression.
+func BuildFTSToILikeExpression(
+	ctx BuildContext,
+	columns []Expression,
+	searchText string,
+	modifier ast.FulltextSearchModifier,
+) (Expression, error) {
```
Comment (qw4990) [inline pkg/expression/fts_to_like.go:170]: I'm not sure whether the current implementation can handle all cases, including some corner and complex cases with special symbols like `<>~*")(...`. For safety, should we check these special symbols at the beginning, and only allow this implementation to handle basic use cases?
We can only support basic use cases initially, and gradually support more cases as users use this feature more deeply.

Comment (qw4990) [reply pkg/expression/fts_to_like.go:170]: What do you think of this? @winoros @AilinKid

File: pkg/expression/fts_to_like.go [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+			Value:   types.NewIntDatum(0),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}
+	}
+
+	if searchText == "" {
+		return zeroIntConst(), nil
+	}
+
+	if modifier.IsBooleanMode() {
```
Comment (qw4990) [inline pkg/expression/fts_to_like.go:195]: Since this function is large, could we change its structure to this to make it easier to review:
```
if modifer.IsBooleanMode() {...}
else if modifer.IsNaturalLanguageMode() {...}
else {... } // doesn't support
```

File: pkg/planner/core/expression_rewriter.go [hunk 7]
```diff
@@ -2362,6 +2364,247 @@ func (er *expressionRewriter) patternLikeOrIlikeToExpression(v *ast.PatternLikeO
...
+		}
+		colName := name.OrigColName
+		if colName.L == "" {
+			colName = name.ColName
+		}
+		if !tableHasPublicFTSIndexOnColumn(tblInfo, colName.L) {
+			return
+		}
+	}
+	sessVars.StmtCtx.AlternativeLogicalPlanHasFTSWithTiFlash = true
```
Comment (AilinKid) [inline pkg/planner/core/expression_rewriter.go]: AlternativeLogicalPlanHasFTSWithTiFlash serves like a signal to trigger the fts-native round, which only be set to true once we found a match against func could be successfully resolved here, while once another different place occur a match-against couldn‘t be resolved here, we should not trigger that round yet. (meaningless another around)

Comment (AilinKid) [reply pkg/planner/core/expression_rewriter.go]: in a word: only set it true when all places resolution are safe

**Comment by AilinKid** [comment]
Non-blocking: if the intent here is only to align selectivity for single-column `MATCH(col)` predicates, that seems reasonable — multi-column `MATCH(title, body)` would require cross-column OR/AND estimation and may not be accurate enough to justify the extra complexity. But if that is the design, I think the code/comment should say so explicitly. Right now `BuildFTSToILikeExpressionFromBuiltin()` still rewrites multi-column `MATCH`, while `GetSelectivityByFilter()` immediately declines filters referencing more than one column, so the native and LIKE rounds still end up using different costing paths for multi-column cases. It may be clearer either to restrict this substitution to single-column `MATCH` only, or to make the multi-column fallback explicit in the comment.

File: pkg/expression/fts_to_like.go (hunk 5 above)
Comment (qw4990) [reply pkg/expression/fts_to_like.go:170]: MySQL's tokenizer seems complex; different modes have different behaviors. Below is some examples of natural language mode:
Image

My current idea is that we can narrow down the scope of this feature for safety and broaden it as needed in the future. And covering some basic cases seems enough for our initial users.
My thought is that:
1.  For `bool mode`, only support `word` `-word` `+word`, word only contains letters and numbers, can't contain special symbols like '-', for example, `xxx +yyy -zzz` is valid, `xx-yy` is invalid.
2. For `natural language mode`, it only supports `word`, it can't contain any special symbols; `xxx yyy zzz` is valid, `x-x y+y` is invalid.

Not sure whether this can cover our initial users' cases. @terry1purcell

Comment (AilinKid) [reply pkg/expression/fts_to_like.go:170]: make sense, from simple terms to complex ones

File: pkg/sessionctx/stmtctx/stmtctx.go [hunk 8]
```diff
@@ -486,6 +486,17 @@ type StatementContext struct {
...
+	// first build round when alternative logical plans are enabled. When true,
+	// the expression rewriter converts MATCH...AGAINST to LIKE predicates
+	// (predicate contexts only) instead of the native FTSMysqlMatchAgainst builtin.
+	AlternativeLogicalPlanFTSLikeFallback bool
+	// AlternativeLogicalPlanHasFTSWithTiFlash is set during the first (ILIKE)
+	// build round when a MATCH...AGAINST expression is encountered AND the
+	// matched columns' table has TiFlash replicas. This triggers the "fts-native"
+	// alternative round so the native FTS builtin (pushed to TiFlash) can compete
+	// on cost against the ILIKE plan.
+	AlternativeLogicalPlanHasFTSWithTiFlash bool
```
Comment (qw4990) [inline pkg/sessionctx/stmtctx/stmtctx.go]: We have now added 2 flags for this feature, which is a little complex. I think we can simplify it to just use 1 flag.
If `Alternative` is enabled, this PR's behavior is:
1. in the first round, set `LikeFallback` to true and generate an ILIKE TiDB plan (and set `FTSWithTiFlash` to true)
2. in the second round, `FTSWithTiFlash` is true, so generate a TiFlash plan.

We could simplify this to just use one flag, only use `LikeFallback`:
1. in the first round, generate a TiFlash plan by default (just the same default behavior if `Alternative` is disabled)
2. in the second round, set `LikeFallback` to true and generate the ILIKE TiDB plan

The root cause of why we need 2 flags here is that we changed the default behavior of handling FTS queries. The prior default plan (`Alternative` is disabled) is the TiFlash plan.
But when `Alternative` is enabled, the default plan becomes the TiDB plan (generated in the first round).
We can keep the prior default behavior and simplify these 2 flags to 1.

File: pkg/expression/fts_to_like.go (hunk 5 above)
Comment (terry1purcell) [reply pkg/expression/fts_to_like.go:170]: I agree with the concerns — the current rewrite quietly handles things like xx-yy as a single token while MySQL would tokenize it into two words. Therefore - we should narrow the scope. 

Next step: I will add a pre-flight validator that runs in BuildFTSToILikeExpression (so both the planner rewrite and selectivity substitution see it) and rejects anything outside the strict subset (boolean: word/+word/-word, alphanumeric only; natural-language: alphanumeric words only). On rejection, the planner side in matchAgainstToExpression should skip the LIKE fallback and emit the native FTSMysqlMatchAgainst builtin instead — that way TiFlash FTS still serves valid queries correctly,  while queries with no FTS index will surface the existing "FTS index required" error rather than silently producing wrong LIKE results.

Comment (terry1purcell) [reply pkg/expression/fts_to_like.go:170]: Implemented

**Review by qw4990 — APPROVED** [review]

**Comment by terry1purcell** [comment]
/hold

I found a problem in the cost competition

**Review by AilinKid — APPROVED** [review]

**Review by Benjamin2037 — COMMENTED** [review]
LGTM

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by windtalker — APPROVED** [review]
lgtm
